LFS-358: Log out process should be more user friendly

OrdiNeu opened the pull request:
[Jira link](https://phenotips.atlassian.net/browse/LFS-358)

_To test_:  The logged in user's avatar is displayed at the top right, where the old log out button was. Change password was moved to there and has been improved with better responsiveness. The log out button should log the currently signed-in user out. 

In addition, try creating a username with multiple names (I.e. Mary Sue). The avatar's initials should change accordingly.

OrdiNeu committed:
LFS-806: Replace the logout button at the top with user avatar and
dropdown with one logout entry

OrdiNeu committed:
LFS-358 / LFS-806: Replace the logout button at the top with user avatar
and dropdown with one logout entry

Remove debugging code

OrdiNeu committed:
LFS-358 / LFS-807: Add "Change password" entry with link to
admin/userlist for changing the password

OrdiNeu committed:
LFS-358 / LFS-808: Display the password change modal on top of the
current page

Fix some bugs with the password change modal

OrdiNeu committed:
LFS-358 / LFS-806: Display the password change modal on top of the
current page

Better inform the user of errors.

OrdiNeu commented:
Found a bug when logging in with a user with only one name (e.g. admin), fix incoming

OrdiNeu committed:
LFS-358: Log out process should be more user friendly

Fix the popup menu being invisible on the Users page

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/AdminNavbarLinks.jsx:
```diff
@@ -10,46 +10,145 @@
+
+  // TODO: Should we make the username accessible to other components (e.g. via a context)?
+  useEffect(() => {
+    fetch("http://localhost:8080/system/sling/info.sessionInfo.json")
```
Why full URL? This explains why I got a `?` instead of an `A` when logging in as `admin`:
![image](https://user-images.githubusercontent.com/651980/103700990-473dd500-4f73-11eb-9c35-488572e1f822.png)

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/AdminNavbarLinks.jsx:
```diff
@@ -10,46 +10,145 @@
+      </Hidden>
+      <Hidden mdUp implementation="css">
+        <Link href={"/system/sling/logout"}>
+          <IconButton
```
The icons mentioned in the specs are missing

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/AdminNavbarLinks.jsx:
```diff
@@ -10,46 +10,145 @@
+        </IconButton>
+      </Hidden>
+      <Hidden mdUp implementation="css">
+        <Link href={"/system/sling/logout"}>
```
"Logout" looks like a link while "Change password" looks like plain text. They should both look the same - black text.

![image](https://user-images.githubusercontent.com/651980/103700455-79026c00-4f72-11eb-8fc4-7da44b0f5e87.png)

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/AdminNavbarLinks.jsx:
```diff
@@ -10,46 +10,145 @@
+        open={pwdResetSuccessSnackbarOpen}
+        autoHideDuration={6000}
+        onClose={() => setPwdResetSuccessSnackbarOpen(false)}
+        message="Password successfully changed!"
```
```suggestion
        message="Password successfully changed"
```

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/AdminNavbarLinks.jsx:
```diff
@@ -10,46 +10,145 @@
+  const avatarRef = useRef();
+  const headerRef = useRef();
+
+  // TODO: Should we make the username accessible to other components (e.g. via a context)?
```
Yes we should, but it might be out of scope for this PR.

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/AdminNavbarLinks.jsx:
```diff
@@ -10,46 +10,145 @@
+    let usernameSplit = username.split(" ");
+    // First name and either the first middle name, or the last name (depending on the number of spaces)
+    initials = usernameSplit[0]?.[0] + (usernameSplit[1]?.[0] || "")
+      // If there's three names or more, include the last name
```
I'm not sure 3 letters will fit that well in the avatar. Wdyt about using only first and last?

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/AdminNavbarLinks.jsx:
```diff
@@ -10,46 +10,145 @@
-          <Hidden mdUp implementation="css">
+          ref={avatarRef}
+          >
+          <Avatar>{initials}</Avatar>
```
The default avatar looks a bit sad (white on gray). We should add a tough of color. Can we use info blue for now, like we do in the administration > user list?

OrdiNeu committed:
LFS-358: Apply code review suggestions by @marta-

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/AdminNavbarLinks.jsx:
```diff
@@ -10,46 +10,151 @@
-import { Button, Hidden, IconButton, withStyles } from "@material-ui/core";
-// @material-ui/icons
-import ExitToApp from "@material-ui/icons/ExitToApp";
+import { Avatar, ClickAwayListener, Grow, Hidden, IconButton, Link, ListItemIcon, ListItemText, MenuList, MenuItem, Paper, Popper, Snackbar, withStyles } from "@material-ui/core";
```
I find long `import` lists easier to read and subsequent changes easier to see if the imported entities are listed one per line: 

```suggestion
import {
  Avatar,
  ClickAwayListener,
  Grow,
  Hidden,
  IconButton,
  Link,
  ListItemIcon,
  ListItemText,
  MenuList,
  MenuItem,
  Paper,
  Popper,
  Snackbar,
  withStyles
} from "@material-ui/core";
```

OrdiNeu replied:
Applied in `1ea0029` :heavy_check_mark:

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/AdminNavbarLinks.jsx:
```diff
@@ -31,7 +33,7 @@ function HeaderLinks (props) {
   // TODO: Should we make the username accessible to other components (e.g. via a context)?
   useEffect(() => {
-    fetch("http://localhost:8080/system/sling/info.sessionInfo.json")
+    fetch("/system/sling/info.sessionInfo.json")
```
Does `fetchWithRelogin` make sense here?

OrdiNeu replied:
Applied in `1ea0029` :heavy_check_mark: 

I'm not 100% sure it makes sense (they shouldn't be capable of loading this page without being signed in) but it's better to be defensive than not :heavy_check_mark:

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/AdminNavbarLinks.jsx:
```diff
@@ -10,46 +10,151 @@
+        </IconButton>
+      </Hidden>
+      <Hidden mdUp implementation="css">
+        <Link href={"/system/sling/logout"}>
```
I think it would be nice to have the icons in the hidden sidebar as well, for consistency with the other entries.

Also, is there any way we could avoid duplicating the code that generates these entries?

![image](https://user-images.githubusercontent.com/651980/103709153-f5e81280-4f7f-11eb-8987-5ef7fce4a585.png)

OrdiNeu replied:
Applied in `1ea0029` :heavy_check_mark: 

I've also just moved all of the styling to copy everything from the sidebar (which was a bit of a refactor at the same time). This means that `headerLinksStyle.jsx` has been completely removed and merged with `sidebarStyle.jsx`.

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/AdminNavbarLinks.jsx:
```diff
@@ -10,46 +10,151 @@
+                    </ListItemIcon>
+                    <ListItemText primary="Change password" />
+                  </MenuItem>
+                  <Link href={"/system/sling/logout"} color="inherit">
```
"Log out" is underlined on hover. Can we get rid of the underline?

![image](https://user-images.githubusercontent.com/651980/103709196-09937900-4f80-11eb-90bf-e4018ba72f6c.png)

OrdiNeu replied:
Fixed in `d7802ac` :heavy_check_mark:

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/headerLinksStyle.jsx:
```diff
@@ -50,7 +50,13 @@ const headerLinksStyle = theme => ({
+    zIndex: "1200"
+  },
+  avatar: {
+    backgroundColor: theme.palette.primary.main
```
I had suggested light blue background like in the Administration > Users list, but I'm not against `theme.palette.primary` for now. I think we'll need to come up with some coloring conventions and stick to them. @sashaandjic do you have an opinion?

![image](https://user-images.githubusercontent.com/651980/103709806-68a5bd80-4f81-11eb-9c23-5ab45e6eb302.png)

Note: the Administration > Users list avatars will need to change anyway so that each avatar gets a different color, so using that specific shade of blue is not really important.

marta- replied:
Let's keep the `primary` blue for now and we'll decide on avatar coloring later as a separate task.

marta- reviewed (changes requested):
Looks good and works well, just some more minor polishing to be done before approval.

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/AdminNavbarLinks.jsx:
```diff
@@ -10,46 +10,151 @@
+        name={username}
+        requireOldPassword
+        />
+      <Snackbar
```
Can we make this `success` green instead of the default black?

OrdiNeu replied:
Implemented in `56de33a5` :heavy_check_mark:

sashaandjic commented:
Tooltip style differs from other tootips.
![image](https://user-images.githubusercontent.com/23243244/103715759-73ffe580-4f8f-11eb-8a12-6a0a765d5010.png)
![image](https://user-images.githubusercontent.com/23243244/103715778-7a8e5d00-4f8f-11eb-96ad-fb00c8faec47.png)
![image](https://user-images.githubusercontent.com/23243244/103715800-811cd480-4f8f-11eb-88e4-be9f4f6372d7.png)

Also, rather than having "Log out" in a tooltip, I would suggest either displaying the user name like in GMAil or "Your profile and settings" like in JIRA or ...
![image](https://user-images.githubusercontent.com/23243244/103716000-fc7e8600-4f8f-11eb-92a2-bfa2da9fcbc9.png)
![image](https://user-images.githubusercontent.com/23243244/103715978-f2f51e00-4f8f-11eb-86fa-261ea4e01b8f.png)

sashaandjic commented:
Administration and "Cards 4 CaRe ..." should be oinned down
![image](https://user-images.githubusercontent.com/23243244/103716544-56cc1680-4f91-11eb-9c40-3e5760c63152.png)
![image](https://user-images.githubusercontent.com/23243244/103716551-5af83400-4f91-11eb-815e-faa4ae0ee407.png)

Alignment and spacing are a bit awkward. Would left alignment of "Search", "Log out" and "Change password" look better?

marta- commented:
> Tooltip style differs from other tootips.
> Also, rather than having "Log out" in a tooltip, I would suggest either displaying the user name like in GMAil or "Your profile and settings" like in JIRA

Right, good catch. The "Log out" title is a leftover from the previous (log out only) icon. Suggested changes:
* remove the `title` attribute and use an actual `Tooltip` component
* the tooltip should be the username

Once we actually implement user profiles ([LFS-820](https://phenotips.atlassian.net/browse/LFS-820)) we can switch the tooltip to "Your profile and settings" or "\<username\>'s profile and settings" ([LFS-810](https://phenotips.atlassian.net/browse/LFS-810)), but not in this PR.

marta- commented:
> Alignment and spacing are a bit awkward. Would left alignment of "Search", "Log out" and "Change password" look better?

Agreed. In one of my review comments I had suggested adding icons to Log out and Change password (and making the alignment more consistent) in the hidden sidebar.

> Administration and "Cards 4 CaRe ..." should be pinned down.

Agreed, but this is out of scope and should be reported as a separate task.

OrdiNeu committed:
LFS-358: Log out process should be more user friendly

Apply code review suggestions by @marta-
Merge the headerLinksStyle with sidebarStyle in order to use its styling

OrdiNeu committed:
LFS-358: Log out process should be more user friendly

Remove old placeholder code

OrdiNeu committed:
LFS-358: Log out process should be more user friendly

Switch the Link to an onClick to remove the underlining on hover

OrdiNeu committed:
LFS-358: Log out process should be more user friendly

Fix the broken Snackbar styling

OrdiNeu committed:
LFS-358: Log out process should be more user friendly

Apply code review comments by @marta-

OrdiNeu commented:
> > Tooltip style differs from other tootips.
>
>     * remove the `title` attribute and use an actual `Tooltip` component
> 
>     * the tooltip should be the username
> 
>     * also remove the `title` attributes from the menu entries (Log out and Change password) since their text is self explanatory

Fixed in `ce89871` :heavy_check_mark:

sashaandjic commented:
Looks awesome! Just one teeny tiny thing if it is easy to do: add more room between icon and text.
![image](https://user-images.githubusercontent.com/23243244/103941629-23a69600-50fd-11eb-929c-039d8be6ef20.png)
I will approve this PR and mark it as tested without this cherry on top.

sashaandjic reviewed (approved):
tested

marta- commented:
> add more room between icon and text.

I think this really needs to be done before merge

marta- commented on modules/homepage/src/main/frontend/src/themePage/Navbars/AdminNavbarLinks.jsx:
```diff
@@ -10,50 +10,157 @@
+
+  const menuItems = <MenuList role="menu">
+    <MenuItem onClick={() => setPasswordDialogOpen(true)} className={expand ? "" : classes.itemLink}>
+      <VpnKeyIcon className={expand ? "" : classNames(classes.itemIcon, classes.whiteFont)}/>
```
Would the spacing problem in the dropdown be solved if we wrap the icon in a `ListItemIcon` when `expand` is true?

OrdiNeu committed:
LFS-358: Log out process should be more user friendly

Fix the spacing between the icon and text

sashaandjic commented:
retested after the last commit - just in case - looks good!
![image](https://user-images.githubusercontent.com/23243244/103952239-10e88d00-510e-11eb-902f-c919561370a1.png)

sdumitriu reviewed (commented):
Bug:
- open the dashboard
- change the password
- go to admin/users list -> change password dialog shows up instead of the user list (bug 1)
- cancel the change password dialog -> you're taken back to the previous page (bug 2)

Also, on the admin/users list page, the two search bars at the top are too similar and confusing, looks like a bug; the user filtering should be restyled.

sashaandjic commented:
Confirming Sergiu's findings.
Bug #1: if I do the refresh after a successful password change, all looks good. If not I get prompted for the password change again. If I cancel and open user list, all is fine again.
Bug #2: is a known issue.
